main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -172,20 +177,48 @@ object TypeReduction {
               case (clazz, tpe) => Type.getFlixType(clazz) == tpe
             })
         exactMethod.length match {
-          case 1 => JavaResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmMethod(exactMethod.head), loc))
-          case _ => JavaResolutionResult.AmbiguousMethod(candidateMethods.toList) // 0 corresponds to no exact method, 2 or higher should be impossible in Java
+          case 1 => JavaMethodResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmMethod(exactMethod.head), loc))
+          case _ => JavaMethodResolutionResult.AmbiguousMethod(candidateMethods.toList) // 0 corresponds to no exact method, 2 or higher should be impossible in Java
         }
     }
   }
 
-  /**
+    /**
+     * Helper method to retrieve a constructor given its parameter types and the class.
+     * Returns a JavaConstructorResolutionResult containing either the constructor, a list of candidate constructors or
+     * a ConstructorNotFound object. The working process is similar to retrieveMethod and differs in the selection of candidate constructors.
+     */
+  private def retrieveConstructor(clazz: Class[_], ts: List[Type], loc: SourceLocation)(implicit flix: Flix): JavaConstructorResolutionResult = {
+    val candidateConstructors = clazz.getConstructors.filter(c => isCandidateConstructor(c, ts))
+
+    candidateConstructors.length match {
+      case 0 => JavaConstructorResolutionResult.ConstructorNotFound
+      case 1 =>
+        val tpe = Type.Cst(TypeConstructor.JvmConstructor(candidateConstructors.head), loc)
+        JavaConstructorResolutionResult.Resolved(tpe)
+      case _ =>
+        // Among candidate constructors if there already exists the one with the exact signature,
+        // we should ignore the rest.
+        val exactConstructor = candidateConstructors
+          .filter(c => // Parameter types correspondance with subtyping
+            (c.getParameterTypes zip ts).forall {
+              case (clazz, tpe) => Type.getFlixType(clazz) == tpe
+            })
+        exactConstructor.length match {
+          case 1 => JavaConstructorResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmConstructor(exactConstructor.head), loc))
+          case _ => JavaConstructorResolutionResult.AmbiguousConstructor(candidateConstructors.toList) // 0 corresponds to no exact constructor, 2 or higher should be impossible in Java
```

Comment:
For this and the corresponding "method" case, I suggest we match directly on 0 and then add a default case that gives an `InternalCompilerException`.